e2e-tests/src/test/committee_split.rs

```diff
@@ -0,0 +1,151 @@
+    const BASE: &str = "ws://127.0.0.1";
+    const FIRST_PORT: u32 = 9944;
+
+    let port = FIRST_PORT + index;
+
+    format!("{BASE}:{port}")
+}
+
+async fn disable_validators(indexes: &[u32]) -> anyhow::Result<()> {
+    info!("Disabling {:?} validators", indexes);
+    let mut connections = vec![];
+    for &index in indexes {
+        let validator_seed = get_validator_seed(index);
+        let stash_controller = NodeKeys::from(validator_seed);
+        let controller_key_to_disable = stash_controller.controller;
+        let address = validator_address(index);
+
+        connections.push(SignedConnection::new(&address, controller_key_to_disable).await);
+    }
+
+    set_invalid_keys_for_validator(connections).await
+}
+
+/// Setup finality committee to be of constructed from 3 reserved nodes and 1 non-reserved node.
+/// 7 nodes are responsible for creating blocks (3 reserved, 4 nonreserved). We first kill `validators`
+/// and then check whether finalization stopped and after a time also block-production. This means the killed
+/// validators were in the finality committee (like the should be).
+async fn split_disable(validators: &[u32]) -> anyhow::Result<()> {
+    let config = setup_test();
+    let root_connection = config.create_root_connection().await;
+    let connection = root_connection.as_connection();
+    prepare_test().await?;
+
+    connection.wait_for_n_eras(2, BlockStatus::Finalized).await;
+
+    // For each reserved node disable it and check that block finalization stopped.
+    // To check that we check that at most 2 sessions passed after disabling - we have limit of 20 blocks
+    // created after last finalized block.
+    info!(
+        "Testing if #{:?} reserved validators are in finalization committee",
+        validators
+    );
+    disable_validators(validators).await?;
+    let session_before = connection.get_session(None).await;
+    let block_before = connection
+        .get_best_block()
+        .await?
+        .expect("there should be some block");
+    sleep(SLEEP_DURATION);
+    let session_after = connection.get_session(None).await;
+    let block_after = connection
+        .get_best_block()
+        .await?
+        .expect("there should be some block");
+    assert!(session_before + 2 >= session_after);
+    // at least some blocks were produced after disabling
+    assert!(block_after > block_before + 10);
+
+    Ok(())
+}
+
+#[tokio::test]
+/// Check if reserved node-0 and node-1 are in the finality committee
+async fn split_test_reserved_01() -> anyhow::Result<()> {
+    split_disable(&[0, 1]).await
+}
+
+#[tokio::test]
+/// Check if reserved node-1 and node-2 are in the finality committee
+async fn split_test_reserved_12() -> anyhow::Result<()> {
+    split_disable(&[0, 1]).await
+}
+
+#[tokio::test]
+/// Check if reserved node-0 and node-2 are in the finality committee
+async fn split_test_reserved_02() -> anyhow::Result<()> {
+    split_disable(&[0, 1]).await
```

Comment:
I think you forgot to adjust the numbers after copying. :P But also, relating this to my other comment about the meaning of these tests – what exactly are you trying to check here?

Reply:
facepalm :P Here I am trying to check whether all of [node-0, node-1, node-2] plus one non-reserved node (hard to pinpoint which exactly in the tests) are actually in the finality committee.

Reply:
Hm, but these are in separate tests, so you are not really checking them together anyway... I guess running these three tests in a loop in a single function is not an option?

It really feels weird when these are separate tests, since they seem to be checking very specific combinations in a very specific setting, while covering all cases but _only_ in this setting. This looks to me as if something is going wrong. I would respect having one test with 2/3 nodes picked randomly more, but adding randomness is always a bit of a problem... :/ If you don't have a better idea feel free to leave it as-is (after fixing the copy problem ofc ;p), but maybe you can figure something reasonable out.